### ISSUE TITLE
HP Fortify: URL::base should not blindly trust `$_SERVER['HTTP_HOST']`

### ISSUE THREAD

**Issue by enov** [kohana/core#613]
This is a security issue related to kohana/kohana#74
### How to reproduce:
- Set up or enable the Apache webserver's default virtual host, without a `ServerName` that will respond to the requests with hosts not defined in other available/enabled sites.
- Install a standard Kohana application within the host's DocumentRoot, most probably `/var/www`
- In `Controller_Welcome` alter the index action's response to

``` php
$this->response->body(URL::site('test', 'https'));
```
- Using a browser extension to modify headers (in my case I used "Modify Headers" chrome extension) modify a request header `Host` to `<iframe src="about:blank" onload="alert('XSS')">`
- Go to `http://localhost` or a locally predefined development URL

You will see "XSS" popping up.

So basically we should [not blindly trust `$_SERVER['HTTP_HOST']`](https://github.com/kohana/core/blob/3.3/master/classes/Kohana/URL.php#L86).

**Comment by enov** [comment]
### Affected Systems:
- A webserver config that can accept a _fall-back_ or _catch-all_ server names
- A standard Kohana 3.3.3.1 installation
  - with relative `base_url` defined in `boostrap.php` (as opposed to an absolute `base_url`s, as in:

``` php
Kohana::init(array(
    'base_url'   => '/kohana/',
));
```
- With direct or indirect calls to `URL::base($protocol)` with `$protocol` defined

All above conditions should be met.

**Comment by enov** [comment]
@acoulton what do you think about this? I suggest to add `allowed_hosts` pattern array to init as a fallback to a relative `base_url` like this:

``` php
Kohana::init(array(
    'base_url'   => '/kohana/',
    'allowed_hosts'   => array(
        'example.com',
        '*.another-example.com',
    ),
));
```

somewhat similar to Symfony's [setTrustedHosts](http://api.symfony.com/2.7/Symfony/Component/HttpFoundation/Request.html#method_setTrustedHosts), also [more here](http://symfony.com/doc/current/reference/configuration/framework.html#trusted-hosts).

**Comment by acoulton** [comment]
@enov that seems sensible. Presumably since this is a security fix we should do this as a bugfix in 3.3 even though it'll be breaking?

**Comment by rjd22** [comment]
@enov sounds like a good plan. I recommend using the same a symfony so `trusted_hosts`.

@acoulton you have 2 options here. If it is not set ignore it and only check it if it's set. Another option would throw an exception to force people to secure their applications. But normally on production you make sure your application has `ServerName` set.

**Comment by enov** [comment]
@acoulton @rjd22 I am sorry I could not write back anytime earlier.

Yes, since this is a security fix, it should be done in 3.3 even if it breaks, probably raising an exception on `URL::base($protocol)` call when `base_url` is relative and `trusted_hosts` unset. For those who have their `base_url` absolute, it should not break.

**Comment by acoulton** [comment]
@enov sounds good.

**Comment by enov** [comment]
@acoulton @rjd22 

I am progressing on this very slowly, as I am in a trade show, offline most of the time.

I found that there exist already similar variables on the core class, [Kohana::$hostnames](https://github.com/kohana/core/blob/3.3/master/classes/Kohana/Core.php#L72) and [Kohana::$server_name](https://github.com/kohana/core/blob/3.3/master/classes/Kohana/Core.php#L67). It seems to me that these are unused. The commits https://github.com/kohana/core/commit/584bdca533a593cbec45a9c2eabfb3ae56c36acb and https://github.com/kohana/core/commit/2fd06a1d8ab8901b23c2d322c296e2a8d393f71a does not show the intent of the author, however parent commits suggest that it _might_ have been intended to differentiate between and internal/external requests (just speculating).

Now, I wonder, would you like me to still use the variable `trusted_hosts` or should I just use `$hostnames` as it is already available?

**Comment by rjd22** [comment]
@enov I would go for trusted_hosts. I rather make it configurable and explicit than automagic :)

**Comment by rjd22** [comment]
Oops wrong button

**Comment by acoulton** [comment]
@enov hmm. I think we should use trusted_hosts and deprecate those static variables.

**Comment by enov** [comment]
and I was thinking to adopt those little poor orphaned variables :crying_cat_face:

**Comment by acoulton** [comment]
In general I think the more we use config and the less we use static
variables, the better...

On 5 June 2015 at 13:34, Samuel Demirdjian notifications@github.com wrote:

> and I was thinking to adopt those little poor orphaned variables [image:
> :crying_cat_face:]
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/kohana/core/issues/613#issuecomment-109281018.

## 

Andrew Coulton
Founder
inGenerator Ltd

Follow us on Twitter @inGenerator
Phone us on 0131 510 0271

inGenerator Ltd is a company registered in Scotland (no SC435161) with its
registered office at 8 Craigleith Hill Row, Edinburgh EH4 2JX.

**Comment by enov** [comment]
I agree @acoulton , but what about consistency in this case? `base_url` in bootstrap and `trusted_hosts` in config?

**Comment by acoulton** [comment]
mmm I see the problem. I still think make it config, and then in a future major release we should put base_url into config too. Most of my apps have a whole chunk of the bootstrap dedicated to loading stuff from config and setting it into the static variables, it's very dull.
